File: Python/search-algorithms/recursive_binary_search.py
"""
Recursive Binary search
"""

def recursive_binary_search(list, target):
    if len(list) == 0:
        return None
    
    middle = (len(list) - 1) //2

    if list[middle] == target:
            return True
    if list[middle] < target:
        return recursive_binary_search(list[middle + 1:], target)
    elif list[middle] > target:
            return recursive_binary_search(list[:middle - 1], target)
    return None



def verify(result, target):
    if result:
        print(f"Target {target} is present")
    else:
        print(f"Target {target} not found!")


if __name__ == "__main__":
    # driver code
    numbers = [i for i in range(1, 10)]
    result = recursive_binary_search(numbers, 7)
    verify(result, 7)

    result = recursive_binary_search(numbers, 10)
    verify(result, 10)

    result = recursive_binary_search(numbers, 2)
    verify(result, 2)

    array = ['peter', 'james', 'john', 'niza']
    result = recursive_binary_search(array, 'niza')
    verify(result, 'niza')

    result = recursive_binary_search(array, 'judy')
    verify(result, 'judy')

File: C/search-algorithms/recursive_binary_search.c
#include <stdio.h>
#include "search.h"
#include <math.h>
#include <string.h>

/*
 * entry point
 */
int main()
{
    printf("Recursive Binary Search\n");
    int numbers[11] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    bool result;

    int length = arr_length(numbers);
    int low = 0;
    int high = length - 1;

    result = recursive_binary_search(numbers, 7, low, high);
    verify_binary(result, 7);

    result = recursive_binary_search(numbers, 10, low, high);
    verify_binary(result, 10);

    result = recursive_binary_search(numbers, 0, low, high);
    verify_binary(result, 0);

    return 0;
}

/**
 * binary search algorithm
 */
int recursive_binary_search(int array[], int target, int low, int high)
{
    if (low > high) {
        return -1; // Element not found
    }

    int mid = (low + high) / 2;

    if (array[mid] == target) {
        return mid; // Element found at index 'mid'
    } else if (array[mid] > target) {
        return recursive_binary_search(array, target, low, mid - 1); // Search the left half
    } else {
        return recursive_binary_search(array, target, mid + 1, high); // Search the right half
    }
}